<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Patryk Orzechowski, Artur Pańszczyk" />

<meta name="date" content="2021-05-19" />

<title>runibic.knit</title>

<script>// Hide empty <a> tag within highlighted CodeBlock for screen reader accessibility (see https://github.com/jgm/pandoc/issues/6352#issuecomment-626106786) -->
// v0.0.1
// Written by JooYoung Seo (jooyoung@psu.edu) and Atsushi Yasumoto on June 1st, 2020.

document.addEventListener('DOMContentLoaded', function() {
  const codeList = document.getElementsByClassName("sourceCode");
  for (var i = 0; i < codeList.length; i++) {
    var linkList = codeList[i].getElementsByTagName('a');
    for (var j = 0; j < linkList.length; j++) {
      if (linkList[j].innerHTML === "") {
        linkList[j].setAttribute('aria-hidden', 'true');
      }
    }
  }
});
</script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>







<div id="runibic-parallel-unibic-biclustering-algorithm-for-r" class="section level1">
<h1>runibic: parallel UniBic biclustering algorithm for R</h1>
<p>This package contains an updated parallel implementation of UniBic biclustering algorithm for gene expression data [Wang2016]. The algorithm locates trend-preserving biclusters within complex and noisy data and is considered one of the most accurate among biclustering methods.</p>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>Since their first application to gene expression data [Cheng2000] biclustering algorithms have gained much popularity [Eren2012]. The reason for this is the ability of the methods to simultaneously detect valid patterns within the data that include only subset of rows and columns.</p>
<p>In this package we provide an efficient parallel improved implementation of UniBic biclustering algorithm. This state-of-the-art algorithm is said to outperform multiple biclustering methods on both synthetic and genetic datasets. Our major contributions are reimplementing the method into more modern C++11 language as well as multiple improvements in code (memory management, code refactoring etc.)</p>
</div>
<div id="functions" class="section level2">
<h2>Functions</h2>
<p>This package provides the following main functions: * <code>BCUnibic</code>/<code>runibic</code> - parallel UniBic for continuous data * <code>BCUnibicD</code> - parallel UniBic for discrete data</p>
<p>The package provides some additional functions: * <code>pairwiseLCS</code> - calculates Longest Common Subsequence (LCS) between two vectors * <code>calculateLCS</code> - calculates LCSes between all pairs of the input dataset (Fibonacci heap or standard sort) * <code>backtrackLCS</code> - recovers LCS from the dynamic programming matrix * <code>cluster</code> - main part of UniBic algorithm (biclusters seeding and expanding) * <code>unisort</code> - returns matrix of indexes based on the increasing order in each row * <code>discretize</code> - performs discretization using Fibonacci heap (sorting method used originally in UniBic)</p>
</div>
<div id="installation" class="section level2">
<h2>Installation</h2>
<p>The package may be installed as follows:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw">install.packages</span>(<span class="st">&quot;devtools&quot;</span>)</span>
<span id="cb1-2"><a href="#cb1-2"></a>devtools<span class="op">::</span><span class="kw">install_github</span>(<span class="st">&quot;athril/runibic&quot;</span>)</span></code></pre></div>
</div>
<div id="examples" class="section level2">
<h2>Examples</h2>
<div id="synthetic-data" class="section level3">
<h3>Synthetic data</h3>
<p>This example shows the basic usage of runibic on synthetic data. We start with loading the libraries.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="kw">library</span>(runibic)</span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="kw">library</span>(biclust)</span></code></pre></div>
<p>First we prepare a random matrix.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a>test &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rnorm</span>(<span class="dv">1000</span>), <span class="dv">100</span>, <span class="dv">100</span>)</span>
<span id="cb3-2"><a href="#cb3-2"></a>test</span></code></pre></div>
<p>Then, we run UniBic biclustering algorithm on the dataset. We use a <code>Biclust</code> package wrapper.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a>res &lt;-<span class="st"> </span>biclust<span class="op">::</span><span class="kw">biclust</span>(test, <span class="dt">method =</span> <span class="kw">BCUnibic</span>())</span></code></pre></div>
<p>We can inspect the results returned by the method. Let’s see how many biclusters were detected.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>res<span class="op">@</span>Number</span></code></pre></div>
<p>We could also inspect the rows of a specific bicluster, for example the third one.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a><span class="kw">which</span>(res<span class="op">@</span>RowxNumber[,<span class="dv">3</span>])</span></code></pre></div>
<p>Similarly we could check the indexes of columns of the third bicluster.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="kw">which</span>(res<span class="op">@</span>NumberxCol[<span class="dv">3</span>,])</span></code></pre></div>
<p>Visual analyses are very useful in biclustering. Let’s draw the heatmap with the obtained bicluster using <code>drawHeatmap</code> function from <code>biclust</code> package.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="kw">drawHeatmap</span>(test, res, <span class="dv">3</span>)</span></code></pre></div>
<p>We could also present parallel coordinates plot of the bicluster with function <code>parallelCoordinates</code> from <code>biclust</code> package.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a><span class="kw">parallelCoordinates</span>(test, res, <span class="dv">3</span>)</span></code></pre></div>
<p>Similarly the package could be used to find trends in discrete data. We run <code>runiDiscretize</code> function.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a>A &lt;-<span class="st"> </span><span class="kw">runiDiscretize</span>(<span class="kw">replicate</span>(<span class="dv">10</span>, <span class="kw">rnorm</span>(<span class="dv">20</span>)))</span>
<span id="cb10-2"><a href="#cb10-2"></a>A</span></code></pre></div>
<p>Finally, we run the UniBic algorithm dedicated to work with discrete datasets.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a><span class="kw">BCUnibicD</span>(A)</span></code></pre></div>
</div>
<div id="data-provided-by-other-packages" class="section level3">
<h3>Data provided by other packages</h3>
<p>This example presents how to use runibic package on a sample dataset. After loading all required libraries</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a><span class="kw">library</span>(runibic)</span>
<span id="cb12-2"><a href="#cb12-2"></a><span class="kw">library</span>(biclust)</span></code></pre></div>
<p>we apply UniBic biclustering method to the sample yeast dataset from <code>biclust</code> package.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a><span class="kw">data</span>(BicatYeast)</span>
<span id="cb13-2"><a href="#cb13-2"></a>res &lt;-<span class="st"> </span><span class="kw">biclust</span>(<span class="dt">method =</span> <span class="kw">BCUnibic</span>(), BicatYeast)</span></code></pre></div>
<p>The method has found 92 biclusters. Now we are going to analyze the results. We will start with drawing a heatmap of the first found bicluster using <code>drawHeatmap</code> from <code>biclust</code> package.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a><span class="kw">drawHeatmap</span>(BicatYeast, res, <span class="dv">1</span>)</span></code></pre></div>
<p>Then we may check the result drawing the genes from the bicluster using parallel coordinates plot with <code>parallelCoordinates</code> function from <code>biclust</code> package.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a><span class="kw">parallelCoordinates</span>(BicatYeast, res, <span class="dv">1</span>)</span></code></pre></div>
</div>
<div id="summarized-experiment" class="section level3">
<h3>Summarized experiment</h3>
<p>The third application of runibic is using the algorithm on the dataset from an RNA-Seq experiment (<code>SummarizedExperiment</code>). We start with loading necessary packages (<code>runibic</code>, <code>SummarizedExperiment</code>) and load the airway dataset.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a><span class="kw">library</span>(runibic)</span>
<span id="cb16-2"><a href="#cb16-2"></a><span class="kw">library</span>(SummarizedExperiment)</span>
<span id="cb16-3"><a href="#cb16-3"></a><span class="kw">data</span>(airway, <span class="dt">package =</span> <span class="st">&quot;airway&quot;</span>)</span></code></pre></div>
<p>We will take only a subset of the data to check if the method can return any patterns.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a>se &lt;-<span class="st"> </span>airway[<span class="dv">1</span><span class="op">:</span><span class="dv">20</span>,]</span></code></pre></div>
<p>Let’s see if UniBic detects any pattern in the dataset.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a>res &lt;-<span class="st"> </span><span class="kw">runibic</span>(se)</span></code></pre></div>
<p>The result could be visualized using <code>parallelCoordinates</code> method from <code>biclust</code> package.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a><span class="kw">parallelCoordinates</span>(<span class="kw">assays</span>(se)[[<span class="dv">1</span>]], res[[<span class="dv">1</span>]], <span class="dv">2</span>)</span></code></pre></div>
<p>We can also draw a heatmap with the second bicluster.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a><span class="kw">drawHeatmap</span>(<span class="kw">assays</span>(se)[[<span class="dv">1</span>]], res[[<span class="dv">1</span>]], <span class="dv">2</span>)</span></code></pre></div>
</div>
<div id="gene-expression-dataset" class="section level3">
<h3>Gene expression dataset</h3>
<p>Another very useful application of the runibic package is analyzing real dataset obtained from Gene Expression Omnibus (GEO). For this task we will use three packages: <code>getGEO</code> from <code>GEOquery</code> to download the data, <code>GDS2eSet</code> from <code>affy</code> to convert the dataset to <code>ExpressionSet</code> and finally <code>exprs</code> to extract gene expression matrix.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a><span class="kw">library</span>(GEOquery)</span>
<span id="cb21-2"><a href="#cb21-2"></a><span class="kw">library</span>(affy)</span></code></pre></div>
<p>We download dataset with Rat peripheral and brain regions from Gene Expression Omnibus.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a>gse &lt;-<span class="st"> </span>GEOquery<span class="op">::</span><span class="kw">getGEO</span>(<span class="st">&quot;GDS589&quot;</span>, <span class="dt">GSEMatrix =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
<p>Now we convert dataset to ExpressionSet and take subset of the dataset.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a>eset &lt;-<span class="st"> </span>affy<span class="op">::</span><span class="kw">GDS2eSet</span>(gds)</span>
<span id="cb23-2"><a href="#cb23-2"></a>subset &lt;-<span class="st"> </span>affy<span class="op">::</span><span class="kw">exprs</span>(eset)[<span class="dv">1</span><span class="op">:</span><span class="dv">100</span>,]</span></code></pre></div>
<p>We perform analysis on first 100 of genes.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a>res &lt;-<span class="st"> </span><span class="kw">runibic</span>(subset)</span></code></pre></div>
<p>Finally, we draw the heatmap with the first bicluster using <code>drawHeatmap</code> function from <code>biclust</code> package.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1"></a><span class="kw">drawHeatmap</span>(subset, res, <span class="dv">1</span>)</span></code></pre></div>
</div>
<div id="comparing-the-results-of-different-biclustering-algorithms" class="section level3">
<h3>Comparing the results of different biclustering algorithms</h3>
<p>We load package with QUBIC biclustering algorithm for comparison</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1"></a><span class="kw">library</span>(runibic)</span>
<span id="cb26-2"><a href="#cb26-2"></a><span class="kw">library</span>(QUBIC)</span>
<span id="cb26-3"><a href="#cb26-3"></a><span class="kw">data</span>(BicatYeast)</span></code></pre></div>
<p>Now, we perform biclustering using CC, Bimax, Qubic, Plaid and Unibic:</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1"></a>resCC &lt;-<span class="st"> </span>biclust<span class="op">::</span><span class="kw">biclust</span>(BicatYeast, <span class="dt">method =</span> <span class="kw">BCCC</span>())</span>
<span id="cb27-2"><a href="#cb27-2"></a>resBi &lt;-<span class="st"> </span>biclust<span class="op">::</span><span class="kw">biclust</span>(BicatYeast, <span class="dt">method =</span> <span class="kw">BCBimax</span>())</span>
<span id="cb27-3"><a href="#cb27-3"></a>resQub &lt;-<span class="st"> </span>biclust<span class="op">::</span><span class="kw">biclust</span>(BicatYeast, <span class="dt">method =</span> <span class="kw">BCQU</span>())</span>
<span id="cb27-4"><a href="#cb27-4"></a>resPlaid &lt;-<span class="st"> </span>biclust<span class="op">::</span><span class="kw">biclust</span>(BicatYeast, <span class="dt">method =</span> <span class="kw">BCPlaid</span>())</span>
<span id="cb27-5"><a href="#cb27-5"></a>resUni &lt;-<span class="st"> </span>biclust<span class="op">::</span><span class="kw">biclust</span>(BicatYeast, <span class="dt">method =</span> <span class="kw">BCUnibic</span>())</span></code></pre></div>
<p>The results of clustering could be compared using <code>showinfo</code> from <code>QUBIC</code> package.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1"></a>QUBIC<span class="op">::</span><span class="kw">showinfo</span>(BicatYeast, <span class="kw">c</span>(resCC, resBi, resPlaid, resQub, resUni))</span></code></pre></div>
</div>
<div id="finding-the-longest-common-subsequence-between-two-vectors-1" class="section level3">
<h3>Finding the Longest Common Subsequence between two vectors (1)</h3>
<p>The Longest Common Subsequence (LCS) between two vectors is the longest series of the data that is present in both analyzed vectors. Let’s prepare two vectors for the analysis.</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1"></a>A &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">5</span>, <span class="dv">4</span>, <span class="dv">3</span>)</span>
<span id="cb29-2"><a href="#cb29-2"></a>B &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">4</span>)</span></code></pre></div>
<p>You may notice that the values (1,2,4) are contained in both vectors: in vector A: (1,2,x,x,4,x) and in vector B: (x,1,x,2,x,4)</p>
<p>Let’s use the methods provided by the package to calculate the longest common subsequence. We first check which values are common for both vectors using <code>backtrackLCS</code>.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1"></a><span class="kw">backtrackLCS</span>(A,B)</span></code></pre></div>
<p>Then we calculate using dynamic programming the matrix for Longest Common Subsequence with <code>pairwiselLCS</code></p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1"></a><span class="kw">pairwiseLCS</span>(A,B)</span></code></pre></div>
</div>
<div id="finding-the-longest-common-subsequence-between-two-vectors-2" class="section level3">
<h3>Finding the Longest Common Subsequence between two vectors (2)</h3>
<p>In our next example we will find the length of the longest common subsequence within the matrix We start with preparing input matrix.</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1"></a>A &lt;-<span class="st">  </span><span class="kw">matrix</span>(<span class="kw">c</span>(<span class="dv">11</span>, <span class="dv">17</span>, <span class="dv">12</span>, <span class="dv">10</span>, <span class="dv">8</span>, <span class="dv">9</span>, <span class="dv">19</span>, <span class="dv">15</span>, <span class="dv">18</span>, <span class="dv">13</span>, <span class="dv">14</span>, <span class="dv">7</span>, <span class="dv">4</span>, <span class="dv">6</span>, <span class="dv">16</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">1</span>, <span class="dv">5</span>, <span class="dv">20</span>,</span>
<span id="cb32-2"><a href="#cb32-2"></a>    <span class="dv">17</span>, <span class="dv">1</span>, <span class="dv">8</span>, <span class="dv">15</span>, <span class="dv">5</span>, <span class="dv">10</span>, <span class="dv">2</span>, <span class="dv">12</span>, <span class="dv">9</span>, <span class="dv">7</span>, <span class="dv">3</span>, <span class="dv">14</span>, <span class="dv">11</span>, <span class="dv">4</span>, <span class="dv">6</span>, <span class="dv">16</span>, <span class="dv">20</span>, <span class="dv">13</span>, <span class="dv">19</span>, <span class="dv">18</span>,</span>
<span id="cb32-3"><a href="#cb32-3"></a>    <span class="dv">15</span>, <span class="dv">8</span>, <span class="dv">17</span>, <span class="dv">12</span>, <span class="dv">18</span>, <span class="dv">14</span>, <span class="dv">19</span>, <span class="dv">11</span>, <span class="dv">16</span>, <span class="dv">20</span>, <span class="dv">10</span>, <span class="dv">13</span>, <span class="dv">6</span>, <span class="dv">3</span>, <span class="dv">7</span>, <span class="dv">9</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">5</span>, <span class="dv">4</span>,</span>
<span id="cb32-4"><a href="#cb32-4"></a>    <span class="dv">15</span>, <span class="dv">12</span>, <span class="dv">16</span>, <span class="dv">9</span>, <span class="dv">19</span>, <span class="dv">17</span>, <span class="dv">10</span>, <span class="dv">18</span>, <span class="dv">11</span>, <span class="dv">20</span>, <span class="dv">8</span>, <span class="dv">13</span>, <span class="dv">2</span>, <span class="dv">5</span>, <span class="dv">7</span>, <span class="dv">14</span>, <span class="dv">1</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">6</span>,</span>
<span id="cb32-5"><a href="#cb32-5"></a>    <span class="dv">15</span>, <span class="dv">10</span>, <span class="dv">9</span>, <span class="dv">6</span>, <span class="dv">13</span>, <span class="dv">19</span>, <span class="dv">7</span>, <span class="dv">18</span>, <span class="dv">16</span>, <span class="dv">17</span>, <span class="dv">14</span>, <span class="dv">4</span>, <span class="dv">3</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">20</span>, <span class="dv">12</span>, <span class="dv">5</span>, <span class="dv">11</span>, <span class="dv">8</span>,</span>
<span id="cb32-6"><a href="#cb32-6"></a>    <span class="dv">1</span>, <span class="dv">7</span>, <span class="dv">4</span>, <span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">6</span>, <span class="dv">8</span>, <span class="dv">13</span>, <span class="dv">5</span>, <span class="dv">9</span>, <span class="dv">12</span>, <span class="dv">11</span>, <span class="dv">16</span>, <span class="dv">15</span>, <span class="dv">17</span>, <span class="dv">10</span>, <span class="dv">19</span>, <span class="dv">20</span>, <span class="dv">14</span>, <span class="dv">18</span>,</span>
<span id="cb32-7"><a href="#cb32-7"></a>    <span class="dv">10</span>, <span class="dv">5</span>, <span class="dv">3</span>, <span class="dv">9</span>, <span class="dv">2</span>, <span class="dv">11</span>, <span class="dv">6</span>, <span class="dv">13</span>, <span class="dv">8</span>, <span class="dv">1</span>, <span class="dv">7</span>, <span class="dv">4</span>, <span class="dv">16</span>, <span class="dv">14</span>, <span class="dv">15</span>, <span class="dv">12</span>, <span class="dv">18</span>, <span class="dv">17</span>, <span class="dv">20</span>, <span class="dv">19</span>,</span>
<span id="cb32-8"><a href="#cb32-8"></a>    <span class="dv">10</span>, <span class="dv">5</span>, <span class="dv">1</span>, <span class="dv">12</span>, <span class="dv">8</span>, <span class="dv">11</span>, <span class="dv">7</span>, <span class="dv">13</span>, <span class="dv">6</span>, <span class="dv">4</span>, <span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">18</span>, <span class="dv">14</span>, <span class="dv">15</span>, <span class="dv">9</span>, <span class="dv">17</span>, <span class="dv">16</span>, <span class="dv">20</span>, <span class="dv">19</span>,</span>
<span id="cb32-9"><a href="#cb32-9"></a>    <span class="dv">9</span>, <span class="dv">6</span>, <span class="dv">3</span>, <span class="dv">10</span>, <span class="dv">1</span>, <span class="dv">12</span>, <span class="dv">7</span>, <span class="dv">13</span>, <span class="dv">8</span>, <span class="dv">2</span>, <span class="dv">5</span>, <span class="dv">4</span>, <span class="dv">16</span>, <span class="dv">14</span>, <span class="dv">15</span>, <span class="dv">11</span>, <span class="dv">19</span>, <span class="dv">17</span>, <span class="dv">20</span>, <span class="dv">18</span>,</span>
<span id="cb32-10"><a href="#cb32-10"></a>    <span class="dv">12</span>, <span class="dv">8</span>, <span class="dv">1</span>, <span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">11</span>, <span class="dv">4</span>, <span class="dv">14</span>, <span class="dv">9</span>, <span class="dv">7</span>, <span class="dv">10</span>, <span class="dv">5</span>, <span class="dv">16</span>, <span class="dv">13</span>, <span class="dv">15</span>, <span class="dv">6</span>, <span class="dv">18</span>, <span class="dv">17</span>, <span class="dv">20</span>, <span class="dv">19</span>), <span class="dt">nrow =</span> <span class="dv">10</span>, <span class="dt">byrow =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
<p>We sort data in each row with the function <code>unisort</code> provided by the <code>runibic</code> package.</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1"></a><span class="kw">unisort</span>(A)</span></code></pre></div>
<p>Now we calculate the length of LCS between each pair of rows using <code>calculateLCS</code> either using Fibonacci Heap or standard sort.</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1"></a>lcsFib &lt;-<span class="st"> </span><span class="kw">calculateLCS</span>(A)</span>
<span id="cb34-2"><a href="#cb34-2"></a>lcs &lt;-<span class="st"> </span><span class="kw">calculateLCS</span>(A, <span class="dt">useFibHeap=</span><span class="ot">FALSE</span>)</span></code></pre></div>
<p>You may notice that depending on the method chosen the results may differ.</p>
<p>We can check the length of the longest common subsequence (e.g. LCS between rows 6 and 7 is equal to 10).</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1"></a><span class="kw">list</span>(<span class="dt">a =</span> lcs<span class="op">$</span>a[<span class="dv">2</span>], <span class="dt">b =</span> lcs<span class="op">$</span>b[<span class="dv">2</span>], <span class="dt">lclen =</span> lcs<span class="op">$</span>lcslen[<span class="dv">2</span>])</span></code></pre></div>
</div>
<div id="discretizing-a-matrix-using-the-quantile-method-from-unibic" class="section level3">
<h3>Discretizing a matrix using the quantile method from UniBic</h3>
<p>First we create a random matrix</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1"></a>B &lt;-<span class="st"> </span><span class="kw">replicate</span>(<span class="dv">10</span>, <span class="kw">rnorm</span>(<span class="dv">10</span>))</span>
<span id="cb36-2"><a href="#cb36-2"></a>B</span></code></pre></div>
<p>Discretization from UniBic (using original sorting method using Fibonacci Heap) could be applied using <code>runiDiscretize</code> function.</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1"></a><span class="kw">runiDiscretize</span>(B)</span></code></pre></div>
</div>
<div id="running-unibic-algorithm-step-by-step" class="section level3">
<h3>Running UniBic algorithm step by step</h3>
<p>In order to take full advantage of the modularity offered by runibic, we apply UniBic algorithm step by step using standard stable sort. The last example shows how to use cluster function from the package. After creating a matrix we sort values in each row and calculate LCS between all pairs of rows.</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1"></a>A &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">c</span>(<span class="dv">4</span>, <span class="dv">3</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">5</span>, <span class="dv">8</span>, <span class="dv">6</span>, <span class="dv">7</span>, <span class="dv">9</span>, <span class="dv">10</span>, <span class="dv">11</span>, <span class="dv">12</span>), <span class="dt">nrow =</span> <span class="dv">4</span>, <span class="dt">byrow =</span> <span class="ot">TRUE</span>)</span>
<span id="cb38-2"><a href="#cb38-2"></a>iA &lt;-<span class="st"> </span><span class="kw">unisort</span>(A)</span>
<span id="cb38-3"><a href="#cb38-3"></a>lcsResults &lt;-<span class="st"> </span><span class="kw">calculateLCS</span>(A, <span class="dt">useFibHeap=</span><span class="ot">FALSE</span>)</span></code></pre></div>
<p>Finally, we apply <code>cluster</code> function to inspect the results.</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1"></a><span class="kw">cluster</span>(iA, A, lcsResults<span class="op">$</span>lcslen, lcsResults<span class="op">$</span>a, lcsResults<span class="op">$</span>b, <span class="kw">nrow</span>(A), <span class="kw">ncol</span>(A))</span></code></pre></div>
</div>
</div>
<div id="citation" class="section level2">
<h2>Citation</h2>
<p>For the original sequential biclustering algorithm please use the following citation:</p>
<p><strong>Zhenjia Wang, Guojun Li, Robert W. Robinson, Xiuzhen Huang <em>UniBic: Sequential row-based biclustering algorithm for analysis of gene expression data</em> Scientific Reports 6, 2016; 23466, doi: <a href="https://doi:10.1038/srep23466" class="uri">https://doi:10.1038/srep23466</a></strong></p>
<p>If you use in your work this package with parallel version of UniBic please use the following citation:</p>
<p><strong>Patryk Orzechowski, Artur Pańszczyk, Xiuzhen Huang, Jason H Moore; <em>runibic: a Bioconductor package for parallel row-based biclustering of gene expression data</em> Bioinformatics, , bty512, <a href="https://doi.org/10.1093/bioinformatics/bty512">https://doi.org/10.1093/bioinformatics/bty512</a></strong></p>
<p>BibTex entry:</p>
<pre><code>@article{orzechowski2018runibic,
  author = {Orzechowski, Patryk and Pańszczyk, Artur and Huang, Xiuzhen and Moore, Jason H},
  title = {runibic: a Bioconductor package for parallel row-based biclustering of gene expression data},
  journal = {Bioinformatics},
  volume = {},
  number = {},
  pages = {bty512},
  year = {2018},
  doi = {10.1093/bioinformatics/bty512},
  URL = {http://dx.doi.org/10.1093/bioinformatics/bty512},
  eprint = {/oup/backfile/content_public/journal/bioinformatics/pap/10.1093_bioinformatics_bty512/4/bty512.pdf}
}</code></pre>
</div>
<div id="references" class="section level2">
<h2>References</h2>
<ul>
<li>[Cheng2000] Cheng, Yizong, and George M. Church. “Biclustering of expression data.” Ismb. Vol. 8. No. 2000. 2000.</li>
<li>[Eren2012] Eren, Kemal, et al. “A comparative analysis of biclustering algorithms for gene expression data.” Briefings in bioinformatics 14.3, 2012: 279-292.</li>
<li>[Wang2016] Wang, Zhenjia, et al. “UniBic: Sequential row-based biclustering algorithm for analysis of gene expression data.” Scientific reports 6 (2016): 23466.</li>
</ul>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
